decouple database, remove ownership from the Outstation

blytkerchan opened the pull request:
This removes ownership of the database from the Outstation and has the calling code inject the database in stead. The outstation still owns the event buffers.

_**This is a breaking change**_: client code will not only have to be recompiled, but will have to be changed to integrate this.
The change is trivial: create the database along with the outstation. The compiler will indicate where the code needs to be changed -- there's no place for ambiguity.

Together with the previous changes, this opens the door to using different database implementations and/or sharing databases between Outstation instances.

Masters are not affected.

blytkerchan committed:
Remove Database ownership from Outstation

The Outstation class, context, etc. are now passed a pointer to a
IDatabase, which is a now-decoupled interface for the database
implemented by the Database class -- or possibly something else.

The Outstation classes no longer know how to make a Databas, so they no
longer need the template, nor some of the database-specific
configuration parameters -- all of which have been removed as
appropriate.

This de-couples the database implementation from the Outstation
implementation.

jadamcrain commented:
Ok, I'll have to take some time to think about this. It definitely complicates things slightly for my average user. It has the potential to confuse ppl because the methodology for updating measurements is rather circuitous. 

1) Create concrete impl
2) Inject into new outstation instance
3) Use MeasUpdate class to safely create a changeset to be applied on correct thread...

Odd that the user now has direct write access to the database.

blytkerchan commented:
The MeasUpdate class is only necessary because of the way the now-default database is synchronized.

I agree that that warrants some more refactoring, though: part of the database API could be refactored to make it more apparent how measurements are to be injected into the database — a bit like RTIMDB's producer/consumer model (but using ASIO strands for synchronization).

I'll have a look to see if I can improve it a bit.

rlc 

## 

I am Ronald's iPhone and I approve this message (now that I've edited it extensively using autocorrect)

> On Apr 4, 2016, at 12:19, Adam Crain notifications@github.com wrote:
> 
> Ok, I'll have to take some time to think about this. It definitely complicates things slightly for my average user. It has the potential to confuse ppl because the methodology for updating measurements is rather circuitous.
> 
> 1) Create concrete impl
> 2) Inject into new outstation instance
> 3) Use MeasUpdate class to safely create a changeset to be applied on correct thread...
> 
> Odd that the user now has direct write access to the database.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub

blytkerchan commented:
See https://groups.google.com/forum/#!topic/automatak-dnp3/f4kaYP7gFrw for how I propose to tackle this.

blytkerchan commented:
btw: the branch has been rebased to 2.0.x

blytkerchan committed:
Remove secauth examples - no longer relevant

blytkerchan committed:
Remove Modify method from MeasUpdate and Database

This method is not called anywhere in the code and is not the way
MeasUpdate is "supposed" to be used. It appears to be there only for a
specific corner-case for the dotnet bindings, which IMO should be handled
within the dotnet bindings in stead (there is no reason to pollute the
code with it for the benefit of .NET).

This also makes it easier to refactor MeasUpdate, as suggested here:
https://groups.google.com/d/msg/automatak-dnp3/f4kaYP7gFrw/fijxjMI_BQAJ
because there will simply be less code to work with.

It further makes it easier to integrate with a database that doesn't know
about ASIO and doesn't do the call-back thing.

blytkerchan committed:
Refactor MeasUpdate to MeasurementUpdate

This adds the commit() method.
Adding rollback(), if we ever want it, would be a simple matter of
	delete m_changes
but there are  no use-cases for this as yet.

All call sites have been changed. The class was renamed to make that
easier. The file was renamed to follow the class name, as appears to be
the rule for this project. (It's a good rule to have, IMHO).

This makes Meas[urement]Update exception-safe

It still allocates the m_changes member, which is a ChangeSet which in
turn contains an std::vector, which in turn contains instances of
std::function - so plenty of actual and potential allocations to go
around. Allocating m_changes is inevitable because it may survive the
MeasurementUpdate object (because the commit is asynchronous), although we
could wait for it or use some non-allocating queueing mechanism -- but we
can also just leave that for other database implementations to do...

blytkerchan committed:
Build on any branch

blytkerchan committed:
A cleaner way to update flags

As noted by Adam here: http://bit.ly/1RX5iXH and here:
http://bit.ly/1RX5lCT, we need a way to update flags without updating the
value. This code ddoes that.

I've added a no_value_change_tag type that can be passed to the
constructor (conveniently, using no_value_change), which will have the
database ignore whatever value is encoded in the measurements and take
only the flags and the timestamp.

jadamcrain commented:
Fairly out of sync at this point, so going to close pull request as is.
